### DISCUSSION TITLE
TextField and SecureTextField in Alerts

### DISCUSSION THREAD

**Discussion by johankool** [pointfreeco/swift-composable-architecture#1357] [Ideas]
According to [this answer on StackOverflow](https://stackoverflow.com/a/73014212/60488) it is possible to show text fields in alerts since iOS 16. 

```swift
.alert("Login", isPresented: $presentAlert, actions: {
  TextField("Username", text: $username)
  SecureField("Password", text: $password)
  Button("Login", action: {})
  Button("Cancel", role: .cancel, action: {})
}, message: {
  Text("Please enter your username and password.")
})
```

For the AlertState of TCA I suppose that maps to the AlertState.Button? I was trying to think how it would be possible to add support for that, but Action is way overloaded here, also ButtonAction already exist.

Trying to think what this may look like:

```swift
state.alert = AlertState(
  title: TextState("Login"),
  message: TextState("Please enter your username and password."),
  actions: [
    .text(TextState("Username"), text: state.username, action: .send(.usernameChanged)),
    .secureText(TextState("Password"), text: state.password, action: .send(.passwordChanged)),
    .cancel(TextState("Cancel")),
    .default(TextState("Login"), action: .send(.loginButtonTapped))
   ])
```

Does this make sense?

**Comment by stephencelis** [comment]
@johankool Interesting! We'd definitely be down to add support for this!

**Comment by johankool** [comment]
One downside is that this (using `.text` and `.secureText` as I suggested above) doesn't express things like adding view modifiers to have the proper keyboard and autocompletion etc.

```swift
TextField("Username", text: $username)
  .textContentType(.emailAddress)
  .keyboardType(.emailAddress)
```

**Reply by stephencelis** [reply]
It might be possible to design an API around this, but at the same time it might also be a rabbit hole. Instead of throwing everything into state, maybe we should rethink how to bind essential alert state to the view layer in a more flexible way.

**Reply by Jeehut** [reply]
@stephencelis Is there any news here as to how we could make use of `TextField` in alerts in TCA today? I'm on the `prerelease/1.0` branch, so if there's anything in the new navigation APIs that helps with this, I could use that. Or do I have to revert to vanilla SwiftUI for now to add TextFields to my alerts?

**Reply by kamcma** [reply]
[Asked the same question here](https://github.com/pointfreeco/swift-composable-architecture/discussions/2048#discussioncomment-5724309), and sketched out a way to use `PresentationState` and `PresentationAction` in the reducer but fall back to vanilla SwiftUI in the view.

**Reply by stephencelis** [reply]
@Jeehut Using the vanilla APIs should work for this for now. We'd like to look into shipping better helpers for the more vanilla APIs based on @kamcma's discussion, but haven't gotten to it yet. We also have plans on getting `TextFieldState` into `AlertState` eventually, but need more time to explore it all.

**Reply by nmalevich** [reply]
Hi everyone!
Any updates?

**Reply by mbrandonw** [reply]
Hi @nmalevich, there are not any updates. We recommend just using the vanilla SwiftUI APIs.

**Reply by stephencelis** [reply]
@nmalevich We sketched out a possible implementation awhile ago here: https://github.com/pointfreeco/swift-navigation/compare/main...text-field-state

That said we haven't had the time ourselves to bring it across the finish line. If you or someone else would like to give it a shot, we'd welcome a PR!
